Replace modal store with redux view reducer #6784

mikelinden1 opened the pull request:
#### Summary
Added Redux reducer to toggle modal state. Use ToggleModalButtonRedux instead of ToggleModalButton and provide a modalId prop (defined in constants->modalIdentifiers). There is a ModalController component in the NeedsTeam component that displays the modals.

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/6784

#### Checklist
- [x] Ran `make check-style` to check for style errors (required for all pull requests)
- [x] Ran `make test` to ensure unit and component tests passed
- [x] Added or updated unit tests (required for all new features)

mattermod commented:
Thanks @mikelinden1 for the pull request!

 Per the [**CONTRIBUTING.md**](https://github.com/mattermost/platform/blob/master/CONTRIBUTING.md#submitting-a-pull-request) file displayed when you created this pull request, we need to add you to the list of [approved contributors](https://docs.google.com/spreadsheets/d/1NTCeG-iL_VS9bFqtmHSfwETo5f-8MQ7oMDE5IUYJi_Y/pubhtml?gid=0&single=true) for the Mattermost project.

**Please help complete the Mattermost [contribution license agreement](http://www.mattermost.org/mattermost-contributor-agreement/)?**

This is a standard procedure for many open source projects. Your form should be processed within 24 hours and reviewers for your pull request will be able to proceed.

Please let us know if you have any questions.

We are very happy to have you join our growing community! If you're not yet a member, please consider joining our [Contributors community channel](https://pre-release.mattermost.com/core/channels/tickets) to meet other contributors and discuss new opportunities with the core team.

jwilander commented:
@mikelinden1 you've got a few styling errors (which is causing the build to fail). You can see them locally by running `make check-style`

mikelinden1 commented:
@jwilander Sorry about that! I ran check-style on the server repo by mistake. I just pushed the fixes.

hmhealey commented on components/modal_controller/index.js:
```diff
@@ -0,0 +1,26 @@
+import {connect} from 'react-redux';
+import {bindActionCreators} from 'redux';
+
+import {closeModal} from '../../actions/views/modals';
```
Can you change this to use an absolute path? We prefer to use those except in cases like below where you're importing the component into index.js

mikelinden1 replied:
No problem, I just pushed the change.

hmhealey commented on components/toggle_modal_button_redux/index.js:
```diff
@@ -0,0 +1,26 @@
+
+import ModalToggleButtonRedux from './toggle_modal_button_redux.jsx';
+
+import {openModal, closeModal} from '../../actions/views/modals';
```
And here

mikelinden1 replied:
Ditto above.

jwilander commented:
@lindalumitchell would you be able to give this a test pass? It affects the opening/closing of the edit channel header, view channel info, channel notifications, channel invite and delete channel modals

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-052bb8d50741ebe9b.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-052bb8d50741ebe9b

lindalumitchell commented:
@mikelinden1 @jwilander I've been testing a bit, and found one Javascript error:

1. Channel name drop-down > Manage Members
2. Click `Add Members` button

Observed: Add Members modal does not open, Javascript error displays
Expected: Add Members modal opens, no error

Note: There are some important elements I can't verify at the moment because they're broken on master:
- Edit channel purpose from drop-down (https://mattermost.atlassian.net/browse/PLT-8170)
- Edit channel name and URL from drop-down (https://mattermost.atlassian.net/browse/PLT-8170)
- Edit channel notifications from drop-down (https://mattermost.atlassian.net/browse/PLT-8202)

mikelinden1 commented:
Found the problem. Just had to pass the openDialog function as the showInviteModal prop to ChannelMembersModal. @lindalumitchell @jwilander

mikelinden1 commented:
Also, there is an error when opening the Add Members modal which is also present on Master.
`Warning: Failed prop type: Invalid prop `defaultMessage` of type `object` supplied to `SaveButton`, expected `string`.`

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-0a82dfeff9048b3c3.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-0a82dfeff9048b3c3

saturninoabril commented on components/toggle_modal_button_redux/toggle_modal_button_redux.jsx:
```diff
@@ -0,0 +1,91 @@
+        super(props);
+
+        this.show = this.show.bind(this);
+        this.hide = this.hide.bind(this);
```
For coding standards, we prefer not to have binding like these and would just implement below functions like ``show = (e) => {...}``

mikelinden1 commented:
Yes, ToggleModalButtonRedux should replace ToggleModalButton everywhere eventually. I removed the `.bind(this)` from ToggleModalButtonRedux and added tests for ToggleModalButtonRedux, ModalController, Modal Reducer, and Modal Actions. @saturninoabril

lindalumitchell commented:
@mikelinden1 @jwilander Should I do another test pass at this point?

jwilander commented:
Yes please @lindalumitchell, if you can just confirm the one bug fix and we should be good. Thanks!

mattermod commented:
`Setup Test Server` label detected. Spinmint test server created if build succeeds (checks pass and no conflicts with base branch).

mattermod commented:
Spinmint test server created at: http://i-08ef1cc0053679432.spinmint.com

Test Account 1: Email: `test@test.com` | Password: `passwd`

Test Account 2: Email: `test2@test.com` | Password: `passwd`

Instance ID: i-08ef1cc0053679432

lindalumitchell commented:
@mikelinden1 @jwilander I verified the issue has been fixed, and did a little bit of re-testing, and it looks good!

One caveat: I'm still unable to test Edit Header, Edit Purpose, and Rename Channel from the drop-down, because those are still broken on master. So it looks like we'll need to verify those after this PR is merged, once those issues are fixed.

jwilander commented:
Thanks @lindalumitchell !

@saturninoabril if you're happy with the changes from your feedback, please approve and we can merge

mattermod commented:
Spinmint test server destroyed
